pkg/controllermanager/controller/project/project_activity_control.go

```diff
@@ -17,89 +17,199 @@ package project
+		// skip it. This is to prevent unnecessary reconciliations in case of GCM restart.
+		if objMeta.GetCreationTimestamp().Add(time.Hour).UTC().Before(c.clock.Now().UTC()) {
+			return
 		}
 	}
 
-	log.Info("Updating Project's lastActivityTimestamp", "lastActivityTimestamp", timestamp)
-	if err := updateStatus(ctx, r.gardenClient, project, func() {
-		project.Status.LastActivityTimestamp = &timestamp
-	}); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed updating Project's lastActivityTimestamp: %w", err)
+	key := c.getProjectKey(ctx, objMeta.GetNamespace())
+	if key == "" {
+		return
 	}
-	return reconcile.Result{}, nil
+
+	c.projectActivityQueue.Add(key)
 }
 
-func (c *Controller) updateShootActivity(obj interface{}) {
-	key, err := cache.MetaNamespaceKeyFunc(obj)
+func (c *Controller) projectActivityObjectWithLabelUpdate(ctx context.Context, oldObj, newObj interface{}) {
+	oldObjMeta, err := meta.Accessor(oldObj)
+	if err != nil {
+		return
+	}
+	newObjMeta, err := meta.Accessor(newObj)
 	if err != nil {
-		c.log.Error(err, "Couldn't get key for object", "object", obj)
 		return
 	}
-	c.projectShootActivityQueue.Add(key)
+
+	// skip queueing if the object(secret or quota) doesn't have the "referred by a secretbinding" label
+	_, oldObjHasLabel := oldObjMeta.GetLabels()[v1beta1constants.LabelSecretBindingReference]
+	_, newObjHasLabel := newObjMeta.GetLabels()[v1beta1constants.LabelSecretBindingReference]
+
+	if !oldObjHasLabel && !newObjHasLabel {
+		return
+	}
+
+	key := c.getProjectKey(ctx, newObjMeta.GetNamespace())
+	if key == "" {
+		return
+	}
+
+	c.projectActivityQueue.Add(key)
+}
+
+func (c *Controller) projectActivityBackupEntryUpdate(ctx context.Context, oldObj, newObj interface{}) {
+	oldBackupEntry, ok := oldObj.(*gardencorev1beta1.BackupEntry)
+	if !ok {
+		return
+	}
+	newBackupEntry, ok := newObj.(*gardencorev1beta1.BackupEntry)
+	if !ok {
+		return
+	}
+
+	if apiequality.Semantic.DeepEqual(newBackupEntry.Spec, oldBackupEntry.Spec) {
+		return
+	}
+
+	key := c.getProjectKey(ctx, newBackupEntry.GetNamespace())
+	if key == "" {
+		return
+	}
+
+	c.projectActivityQueue.Add(key)
+}
+
+func (c *Controller) projectActivityPlantUpdate(ctx context.Context, oldObj, newObj interface{}) {
+	oldPlant, ok := oldObj.(*gardencorev1beta1.Plant)
+	if !ok {
+		return
+	}
+	newPlant, ok := newObj.(*gardencorev1beta1.Plant)
+	if !ok {
+		return
+	}
+
+	if apiequality.Semantic.DeepEqual(newPlant.Spec, oldPlant.Spec) {
+		return
+	}
+
+	key := c.getProjectKey(ctx, newPlant.GetNamespace())
+	if key == "" {
+		return
+	}
+
+	c.projectActivityQueue.Add(key)
+}
+
+func (c *Controller) projectActivityShootUpdate(ctx context.Context, oldObj, newObj interface{}) {
+	oldShoot, ok := oldObj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return
+	}
+	newShoot, ok := newObj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return
+	}
+
+	if apiequality.Semantic.DeepEqual(newShoot.Spec, oldShoot.Spec) {
```

Comment:
Let's only check for a changed `.metadata.generation`, I think this is more precise. This way, you can probably harmonize these functions and use it for various objects (`Plant`, `Shoot`, etc.). You only need the metadata, hence can use `meta.Accessor` again.